### DISCUSSION TITLE
Missing Smart Sensors

### DISCUSSION THREAD

**Discussion by Ed-M72** [BottlecapDave/HomeAssistant-OctopusEnergy#228] [Q&A]
Hi Dave,
as you can probably tell I am not seeing my  accumulative cost sensors for the electricity import/export meters.  I checked my account payload and there is no mention of a smart meter in it.  I used to get values for these sensors so must have been a change in the last 3 months which has broken it.

      "electricity_meter_points": [
        {
          "mpan": "20XXXXXXXXX40",
          "profile_class": 0,
          "consumption_standard": 2900,
          "meters": [
            {
              "serial_number": "1XXXXXXXXX1",
              "registers": [
                {
                  "identifier": "1",
                  "rate": "STANDARD",
                  "is_settlement_register": true
                }
              ]
            }
          ],

It could just be an oddity for me,  if it is then lets leave it as I am hoping to have my meters swapped for SMETS2 ones this year.

**Comment by BottlecapDave** [comment]
Hello and sorry that you're having issues. If you could download the diagnostics associated with your electricity meter in HA and paste the contents (not the file) here then that might give a bit more information. Has this ever worked for you or are you new to the integration? 

Currently the integration only adds previous consumption sensors for meters that are reporting as either a smart import or export meter. Current consumption sensors are only added for users who have an Octopus Home Mini.

**Reply by Ed-M72** [reply]
Hi Dave, yes it was working from around the 20th of December.  It stopped working around 6-8 weeks ago and was added to my list of things to investigate.  Your debug log has a flag saying that it is not a smart meter, where is the code that determines that?  The gas meter is definitely not a smart meter, so that is correct.

{
  "home_assistant": {
    "installation_type": "Home Assistant OS",
    "version": "2023.4.6",
    "dev": false,
    "hassio": true,
    "virtualenv": false,
    "python_version": "3.10.10",
    "docker": true,
    "arch": "aarch64",
    "timezone": "Europe/London",
    "os_name": "Linux",
    "os_version": "6.1.21-v8",
    "supervisor": "2023.04.1",
    "host_os": "Home Assistant OS 10.1",
    "docker_version": "23.0.3",
    "chassis": "embedded",
    "run_as_root": true
  },
  "custom_components": {
    "solcast_solar": {
      "version": "3.0.46",
      "requirements": [
        "aiohttp>=3.6.2",
        "datetime>=4.3",
        "isodate>=0.6.0"
      ]
    },
    "sun2": {
      "version": "2.2.1",
      "requirements": []
    },
    "myenergi": {
      "version": "0.0.23",
      "requirements": [
        "pymyenergi==0.0.27"
      ]
    },
    "localtuya": {
      "version": "5.0.0",
      "requirements": []
    },
    "octopus_energy": {
      "version": "7.0.0",
      "requirements": []
    },
    "hacs": {
      "version": "1.32.1",
      "requirements": [
        "aiogithubapi>=22.10.1"
      ]
    },
    "scheduler": {
      "version": "v0.0.0",
      "requirements": []
    },
    "fujitsu_general_heatpump": {
      "version": "1.0.0",
      "requirements": [
        "pyfujitseu"
      ]
    }
  },
  "integration_manifest": {
    "domain": "octopus_energy",
    "name": "Octopus Energy",
    "codeowners": [
      "@bottlecapdave"
    ],
    "config_flow": true,
    "dependencies": [
      "repairs"
    ],
    "documentation": "https://github.com/BottlecapDave/HomeAssistant-OctopusEnergy/",
    "homekit": {},
    "iot_class": "cloud_polling",
    "issue_tracker": "https://github.com/BottlecapDave/HomeAssistant-OctopusEnergy/issues",
    "ssdp": [],
    "version": "7.0.0",
    "zeroconf": [],
    "is_built_in": false
  },
  "data": {
    "electricity_meter_points": [
      {
        "mpan": "**REDACTED**",
        "meters": [
          {
            "serial_number": "**REDACTED**",
            "is_export": false,
            "is_smart_meter": false,
            "device_id": null,
            "manufacturer": "Secure Sprint 211",
            "model": null,
            "firmware": null
          }
        ],
        "agreements": [
          {
            "valid_from": "2023-04-08T23:00:00+00:00",
            "valid_to": "2024-04-08T23:00:00+00:00",
            "tariff_code": "E-1R-FLUX-IMPORT-23-02-14-H",
            "product_code": "FLUX-IMPORT-23-02-14"
          }
        ]
      },
      {
        "mpan": "**REDACTED**",
        "meters": [
          {
            "serial_number": "**REDACTED**",
            "is_export": false,
            "is_smart_meter": false,
            "device_id": null,
            "manufacturer": "Secure Sprint 211",
            "model": null,
            "firmware": null
          }
        ],
        "agreements": [
          {
            "valid_from": "2023-04-08T23:00:00+00:00",
            "valid_to": null,
            "tariff_code": "E-1R-FLUX-EXPORT-23-02-14-H",
            "product_code": "FLUX-EXPORT-23-02-14"
          }
        ]
      }
    ],
    "gas_meter_points": [
      {
        "mprn": "**REDACTED**",
        "meters": [
          {
            "serial_number": "**REDACTED**",
            "consumption_units": "",
            "is_smart_meter": false,
            "device_id": null,
            "manufacturer": "G4",
            "model": null,
            "firmware": null
          }
        ],
        "agreements": [
          {
            "valid_from": "2021-10-19T23:00:00+00:00",
            "valid_to": "2023-10-19T23:00:00+00:00",
            "tariff_code": "G-1R-OE-FIX-24M-21-09-17-H",
            "product_code": "OE-FIX-24M-21-09-17"
          }
        ]
      }
    ]
  }
}

**Reply by BottlecapDave** [reply]
The code that determines if you have a smart meter can be found in https://github.com/BottlecapDave/HomeAssistant-OctopusEnergy/blob/develop/custom_components/octopus_energy/api_client.py#L221 where the data is from their GraphQL API using the query located at https://github.com/BottlecapDave/HomeAssistant-OctopusEnergy/blob/develop/custom_components/octopus_energy/api_client.py#L22. If both `smartImportElectricityMeter` and `smartExportElectricityMeter` is not present then I deem it not to be a smart meter. I added this logic as I had a couple of reports of people using the integration without smart meters.

**Reply by BottlecapDave** [reply]
Hi @Ed-M72 - it would be good to know if I've made the wrong assumption and there is something else that indicates if a smart meter is present or if it's a configuration issue with OE.

**Reply by Ed-M72** [reply]
Hi Dave, I couldn't quite work out how to get the date out of the Kracken engine via your GraphQL query.  Is there a single way to run it in the GraphQL web app? it looks like that is where you are determining if it is a smart meter.

I can see yesterdays consumption history for both import/export though when I call https://api.octopus.energy/v1/electricity-meter-points/{MPAN}/meters/{SerialNo}/consumption/

Ed

**Reply by BottlecapDave** [reply]
To use the graphql webpoint, you first need to get a an access token. This can be obtained by calling the following graphql query (filling in the api key from your dashboard)

```
mutation {
	obtainKrakenToken(input: { APIKey: "{api_key}" }) {
		token
	}
}
```

You then set the value of `token` as a header of `Authorization: JWT <<TOKEN>>` with the following query

```
query {
  account(accountNumber: "{account_id}") {
    electricityAgreements(active: true) {
			meterPoint {
				mpan
				meters(includeInactive: false) {
          makeAndType
					serialNumber
          makeAndType
          smartExportElectricityMeter {
						deviceId
            manufacturer
            model
            firmwareVersion
					}
          smartImportElectricityMeter {
						deviceId
            manufacturer
            model
            firmwareVersion
					}
				}
				agreements {
					validFrom
					validTo
					tariff {
						...on StandardTariff {
							tariffCode
              productCode
						}
						...on DayNightTariff {
							tariffCode
              productCode
						}
						...on ThreeRateTariff {
							tariffCode
              productCode
						}
						...on HalfHourlyTariff {
							tariffCode
              productCode
						}
            ...on PrepayTariff {
							tariffCode
              productCode
						}
					}
				}
			}
    }
    gasAgreements(active: true) {
			meterPoint {
				mprn
				meters(includeInactive: false) {
					serialNumber
          consumptionUnits
          modelName
          smartGasMeter {
						deviceId
            manufacturer
            model
            firmwareVersion
					}
				}
				agreements {
					validFrom
					validTo
					tariff {
						tariffCode
            productCode
					}
				}
			}
    }
  }
}
```

As mentioned, there is one of two things that are different with your account

1) Your account has been configured incorrectly (which I've seen reports of in other areas) and therefore you need to contact Octopus Energy or
2) There is another value within the second query that indicates that your meter is "smart". This option will need more investigation.

**Reply by Ed-M72** [reply]
Ah brilliant, thanks for the code Dave,

Looks like Octopus set up my meter correctly.  I will phone them to get it corrected.

"smartExportElectricityMeter": null,
"smartImportElectricityMeter": null

**Reply by BottlecapDave** [reply]
I'm not sure if call center staff would be able to help with this. I would recommend [emailing](mailto:hello@octopus.energy?subject=Customer%20API) or asking on their forums. There is another potential field I could use, but it doesn't give me enough information for knowing if exports are supported or that your gas meter is smart.
